File: uri/uri_cpp/matematica/p2129.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <vector>

using namespace std;


int main() {
	int i, k = 1, n, N = 1000000;
	vector <int> fat;
	fat.push_back(1);
	long fatorial = 1, mod = 1000000000;
	for (i = 1; i <= N; i++) {
		fatorial = (fatorial*i) % mod;
		while (fatorial % 10 == 0)
			fatorial /= 10;
		fat.push_back(fatorial % 10);
	}
	while (cin >> n)
		cout << "Instancia " << k++ << endl << fat[n] << endl << endl;

    return 0;
}

File: uri/uri_python/matematica/p2129.py
N = 1000000
fat = [1]
mod = 1000000000
for i in range(1, N + 1):
	fat.append(fat[i-1]*i % mod)
	while fat[i] % 10 == 0:
		fat[i] = int(fat[i]/10)
#for i in range(N):
#	print('i, f: ', i, fat[i])
k = 1
while (True):
	try:
		n = int(input())
		print("Instancia", k)
		print(fat[n] % 10)
		print("")
		k += 1
	except EOFError:
		break

